### PR TITLE
worked on integrand visualization

### PR DESCRIPTION
**Pull request by heikef**
Worked on integrating visualizations for unstructured grids taking the work of @lnw as example. However, the surface pictures look bad and I do not understand why. Now it is time to share and ask for comments.

### CODE REVIEW

**Comment by lnw** [comment]
do you have an example for what goes wrong?

File: src/fgimic/vtkplot.f90 [hunk 1]
```diff
@@ -311,5 +311,83 @@ subroutine write_vtk_vector_unstructuredgrid(fname, grid, pdata, cells)
         call closefd(fd)
     end subroutine
 
+    ! fname: output filename
+    ! grid: grid point coordinates
+    ! pdata: vectors at grid points
```
Comment (lnw) [inline src/fgimic/vtkplot.f90:316]: those are now scalars at grid points

File: src/fgimic/vtkplot.f90 [hunk 2]
```diff
@@ -311,5 +311,83 @@ subroutine write_vtk_vector_unstructuredgrid(fname, grid, pdata, cells)
...
+    ! cell definitions
+    subroutine write_vtk_imagedata_unstructuredgrid(fname, grid, pdata, cells)
```
Comment (lnw) [inline src/fgimic/vtkplot.f90]: I think I'd call this 'write_vtk_scalar_unstructuredgrid'.

File: src/fgimic/jfield.f90 [hunk 3]
```diff
@@ -674,6 +701,66 @@ subroutine get_property(this)
           tmp = (sigma(1) + sigma(2) + sigma(3))/3.0d0
           write(*,*) "shielding constant sigma in ppm= ", tmp
 
+          if(elements_exists) then
+              if (k < 10) then
```
Comment (lnw) [inline src/fgimic/jfield.f90]: that's one level of indentation too much

File: src/fgimic/jfield.f90 [hunk 4]
```diff
@@ -627,12 +635,29 @@ subroutine get_property(this)
...
+        ! read element file: first number in first line is number of lines
+        inquire(FILE='grid.1.ele', EXIST=elements_exists)
+        if(elements_exists) then
+          open(GRIDELE, file='grid.1.ele')
+          ! read(GRIDELE, '(3i4)') ncells, dummy, dummy
+          read(GRIDELE, *) ncells, dummy, dummy
+          ! write(*,*) "ncells ", ncells
+          allocate(cells(4,ncells))
+          do i=1,ncells
+            ! read(GRIDELE,'(5i7)') dummy, cells(1:4, i)
```
Comment (lnw) [inline src/fgimic/jfield.f90]: i'd probably delete a few outcommented lines here and further down

File: src/fgimic/jfield.f90 [hunk 5]
```diff
@@ -674,6 +701,66 @@ subroutine get_property(this)
...
+              call write_vtk_imagedata_unstructuredgrid(trim(filename), this%grid%xdata, intsigma(:,2), cells)
+              write (filename,formatstring) "sigma_zz", k, ".vtu"
+              print *, filename 
+              call write_vtk_imagedata_unstructuredgrid(trim(filename), this%grid%xdata, intsigma(:,3), cells)
+          end if
+        end do !loop atoms
+        write(*,*) ""
+        
+        ! calculate diagonal elements of magnetizabilitz tensor
+        ! for magnetizabilitz no loop over coords is needed
```
Comment (lnw) [inline src/fgimic/jfield.f90]: someone has a German keyboard layout :)

File: src/fgimic/vtkplot.f90 (hunk 1 above)
Comment (heikef) [reply src/fgimic/vtkplot.f90:316]: but you are in the vector file

File: src/fgimic/vtkplot.f90 (hunk 2 above)
Comment (heikef) [reply src/fgimic/vtkplot.f90]: hm, this is a matter of taste. For the structured grid we called it also "image data".

File: src/fgimic/jfield.f90 (hunk 3 above)
Comment (heikef) [reply src/fgimic/jfield.f90]: ok

File: src/fgimic/jfield.f90 (hunk 4 above)
Comment (heikef) [reply src/fgimic/jfield.f90]: Ok. When everything works we can do a clean up.

File: src/fgimic/jfield.f90 (hunk 5 above)
Comment (heikef) [reply src/fgimic/jfield.f90]: Nope, a Norwegian one.

**Comment by heikef** [comment]
@lnw Thanks for commenting and reviewing! Despite of the name of the vtu scalar print file you did not comment further. I assume the modification was correct then. When removing the integral weights the visuals looked also much better.

**Comment by heikef** [comment]
@lnw and @bast when you are ok now with the changes I would suggest to merge and then the decomposition into positive and negative contributions from Rahul @rahulj98 could be added

File: src/fgimic/vtkplot.f90 (hunk 2 above)
Comment (lnw) [reply src/fgimic/vtkplot.f90]: to me, we are printing a scalar field (which also applies to the comment one up).  This is in analogy to the vector field that is printed in the other function that is called 'write_vtk_vector_unstructuredgrid'.

**Comment by lnw** [comment]
I can't comment on the correctness of the equations, such as the factor of 1/3, but I did verify that the printed scalar fields are now smooth.  Apart from that it looks very useful.

File: src/fgimic/vtkplot.f90 (hunk 2 above)
Comment (heikef) [reply src/fgimic/vtkplot.f90]: I changed the name in my last commit.

**Comment by heikef** [comment]
@lnw thank you for the verification. I removed the factor of 1/3 because then one can use larger isovalues for visualization and the resulting plots remain the same. Thus, the factor does not change the visual information (at least for c4h4 and h2o).

**Comment by heikef** [comment]
Cleaned up comments and squashed commits.

**Comment by heikef** [comment]
@bast I would merge now. Then Rauhl could contintue and file a new pull request.

**Pull request merged**